### ISSUE TITLE
Addition and multiplication operations are not available.

### ISSUE THREAD

**Issue by MizunagiKB** [MizunagiKB/gd_cubism#7]
The current GDCubism does not correctly reproduce part drawings with addition and multiplication specified.
Therefore, part drawing with addition and multiplication specified is forcibly switched to the shader that performs normal drawing.

![problem_01](https://github.com/MizunagiKB/gd_cubism/assets/5047878/fd57c4b6-21ef-4f10-a067-ab3e0102aa66)

No effective solution has been found at this time.

- - - -

現在のGDCubismでは加算と乗算指定がされたパーツ描画が正しく再現されません。
そのため加算と乗算指定がされたパーツ描画には通常描画を行うシェーダーに強制的に切り替わる様になっています。

現在有効な解決策が見つかっていません。

**Comment by fnaith** [comment]
Does that mean official demo applies add+mul blending with only one layer, but godot must uses two layers?

**Comment by MizunagiKB** [comment]
The problem is that both Official and GDCubism use the same number of layers, but the drawing results are different.

This problem is mainly due to two factors
* Godot Engine's drawing rules cannot be changed.
* I am not familiar with how shaders work.

So I added the ability to change shaders to the latest commit of 0.1.

* bcf247efc7d495a38b297fc4dd9c256d95ab9512
* e7a8b1792da961a1c23c41786a2843b606ecbb02
* 64864ece11a3c1ae4808a4708ce24f32f3d7ecd8

![ss](https://github.com/MizunagiKB/gd_cubism/assets/5047878/456d5c91-1c97-456f-a587-5e8c22800a9b)

You can edit the shaders directly in the /demo/addons/gd_cubism/res/shader folder, or copy them to another location and set them in your editor.

The most obvious one is shadow rendering, where the following shaders are used.

* 2d_cubism_mask_mul.gdshader

Even if you can't find a solution, I would be very happy to hear what you find.

**Comment by fnaith** [comment]
@MizunagiKB  thanks for your sharing!
Although I am not familiar with godot shader too, but I will build this project and try to find out.
It is good see some talented person works on godot live2d.
Recent Unity Policy changes many things.

**Comment by fnaith** [comment]
the build process is very smooth !
![demo](https://github.com/MizunagiKB/gd_cubism/assets/11421034/a04d8625-bc7b-4055-9cec-70659335a379)

**Comment by MizunagiKB** [comment]
I am happy to build in your environment.

**Comment by MizunagiKB** [comment]
Updated Shader to improve drawing reproducibility.
Some Live2D models may not display well.

* https://github.com/MizunagiKB/gd_cubism/tree/0.2

![ss_mao](https://github.com/MizunagiKB/gd_cubism/assets/5047878/0f834f70-f3d2-4c1a-8329-54987284eaf5)

**Comment by fnaith** [comment]
So cool, I guess it solves a part of problem 👍

**Comment by kuwana-live2d** [comment]
Is this problem perhaps caused by not being able to set a custom blend mode in the shader?
Is the problem related to the following proposals?

[Add more 2D blend modes #470](https://github.com/godotengine/godot-proposals/issues/470)
[Add Custom Blend Mode declarations to the shader language #7058](https://github.com/godotengine/godot-proposals/issues/7058)

**Comment by MizunagiKB** [comment]
> So cool, I guess it solves a part of problem 👍

@fnaith 

I don’t think this method is a panacea, but if the appearance is within an acceptable range, I would like to release it as 0.2.

**Comment by MizunagiKB** [comment]
@kuwana-live2d 

As far as I understand, the current Godot Engine does not allow you to freely specify the blend mode.
Therefore, if you perform addition or multiplication on textures with alpha, you will get unintended rendering results. This Issues(#7) is trying to solve the problem by using Shader tricks without modifying the game engine.

The links you provided seem to be related to the blend mode topic. Maybe if what is mentioned in the links is implemented, the problem will be solved.

However, I think my approach is quite exceptional for Godot Engine.
> This is also related to the fact that Live2D itself uses a unique technique.

Therefore, at the moment, I think it would be best if we could find a solution using just the GDExtension.

**Comment by kuwana-live2d** [comment]
@MizunagiKB 

As far as I can tell, the recommended way to use custom blends in godot is to use [screen texture](https://docs.godotengine.org/en/latest/tutorials/shaders/screen-reading_shaders.html). 

However, according to [the document](https://docs.godotengine.org/en/latest/tutorials/shaders/screen-reading_shaders.html#:~:text=c%3B%0A%7D-,Behind%20the%20scenes,%C2%B6,-While%20this%20seems), to get screen texture multiple times, I think you need to use a lot of back buffer copies.

Since live2d sets the blend method when drawing the mesh, I am thinking that you may need to create a backbuffercopy in the drawing unit of that mesh.

I don't know what the performance impact of this method would be, since we don't normally blend in this manner...

**Comment by MizunagiKB** [comment]
@kuwana-live2d 

Given that the blend mode cannot be changed, in order to achieve the desired composite result, it is necessary to obtain information about the rendering destination and write your own composite formula.

Indeed, it is as you have stated.

In GDCubism, we use sub-viewport instead of screen texture, which is used for overall rendering and mask processing. Therefore, if you obtain the results rendered to the sub-viewport as needed, you should be able to perform more complex calculations.
However, obtaining the rendering results increases the load and complicates the process. Above all, it exceeds my understanding as a layman.

Currently, we are aiming for a reasonable level of reproduction and are combining the following to create a composition that does not destroy the alpha of the rendering destination:

* blend_premul_alpha (mix, add)
* blend_mul + alpha 1.0 (mul)

Additionally, we have made shaders available as properties, enabling users to customize them on their own.

**Comment by kuwana-live2d** [comment]
@MizunagiKB 

I feel that changing sub viewport to back buffer copy(screen texture) would allow you to get the rendering results in the shader, but I am concerned about the performance loss.

I have understood your situation that you are improving the shader code to get as close to correct rendering as possible.

I'll try to find a few more ways you can improve.

**Comment by thynetruly** [comment]
Just notifying this team that there has been a pull request into the godot engine that as of two weeks ago makes available blend mode support to users. The proposal is linked my mr11stein and the PR is available therein.

**Comment by MizunagiKB** [comment]
> Is this the pull request you are referring to?  
https://github.com/godotengine/godot/pull/102366  

The rendering issues we initially encountered have been significantly resolved thanks to the contributions and fixes by everyone involved. Additionally, starting from version **0.9**, we plan to switch to a method that does not rely on `SubViewport` for final output.  

While the situation has changed considerably from the beginning, if it is now possible to set the blend mode, it might prove useful in some cases.  

Thank you for letting us know.

**Comment by erodozer** [comment]
Just want to document what I found so we know what needs to be changed to fix blending for L2D models if ever https://github.com/godotengine/godot/pull/102366 is merged

Godot's [blend modes](https://github.com/godotengine/godot/blob/bf95b62586e31b8a3503f5903d7764d7c52bf2ab/drivers/gles3/storage/texture_storage.cpp#L1459) are

|mode|src color|dst color|src alpha|dst alpha|
|-|-|-|-|-|
|mix|GL_SRC_ALPHA|GL_ONE_MINUS_SRC_ALPHA|GL_ONE|GL_ONE_MINUS_SRC_ALPHA|
|add|GL_ONE|GL_ONE|GL_ONE|GL_ONE|
|mul|GL_DST_COLOR|GL_ZERO|GL_DST_ALPHA|GL_ZERO|

Whereas Live2D's [reference implementation](https://github.com/Live2D/CubismNativeFramework/blob/5-r.4.1/src/Rendering/OpenGL/CubismShader_OpenGLES2.cpp#L425) uses
|mode|src color|dst color|src alpha|dst alpha|
|-|-|-|-|-|
|mix|GL_ONE|GL_ONE_MINUS_SRC_ALPHA|GL_ONE|GL_ONE_MINUS_SRC_ALPHA|
|add|GL_ONE|GL_ONE|GL_ZERO|GL_ONE|
|mul|GL_DST_COLOR|GL_ONE_MINUS_SRC_ALPHA|GL_ZERO|GL_ONE|
